clang/runtime/dpct-rt/include/dpct/sparse_utils.hpp

```diff
@@ -40,13 +40,124 @@ class matrix_info {
   oneapi::mkl::uplo _uplo = oneapi::mkl::uplo::upper;
   oneapi::mkl::index_base _index_base = oneapi::mkl::index_base::zero;
 };
+namespace detail {
+using csrgemm_args_info =
+    std::tuple<oneapi::mkl::transpose, oneapi::mkl::transpose, int, int, int,
+               const std::shared_ptr<matrix_info>, const void *, const int *,
+               const int *, const std::shared_ptr<matrix_info>, const void *,
+               const int *, const int *, const std::shared_ptr<matrix_info>,
+               const int *>;
+struct csrgemm_args_info_hash {
+  size_t operator()(const csrgemm_args_info &args) const {
+    std::stringstream ss;
+    ss << (char)std::get<0>(args) << ":";
+    ss << (char)std::get<1>(args) << ":";
+    ss << std::get<2>(args) << ":";
+    ss << std::get<3>(args) << ":";
+    ss << std::get<4>(args) << ":";
+    ss << std::get<5>(args).get() << ":";
+    ss << std::get<6>(args) << ":";
+    ss << std::get<7>(args) << ":";
+    ss << std::get<8>(args) << ":";
+    ss << std::get<9>(args).get() << ":";
+    ss << std::get<10>(args) << ":";
+    ss << std::get<11>(args) << ":";
+    ss << std::get<12>(args) << ":";
+    ss << std::get<13>(args).get() << ":";
+    ss << std::get<14>(args) << ":";
+    return std::hash<std::string>{}(ss.str());
+  }
+};
+#ifdef __INTEL_MKL__
+template <typename handle_t, typename init_func_t, typename rel_func_t>
+class handle_manager {
+public:
+  handle_manager(sycl::queue *q, init_func_t init_func, rel_func_t rel_func)
+      : _q(q), _init_func(init_func), _rel_func(rel_func) {
+    _h = new handle_t;
+    _init_func(_h);
+  }
+  ~handle_manager() {
+    if constexpr (std::is_same_v<handle_t,
+                                 oneapi::mkl::sparse::matrix_handle_t>) {
+      sycl::event e = _rel_func(*_q, _h, _deps);
+      _q->submit([&](sycl::handler &cgh) {
+        cgh.depends_on(e);
+        cgh.host_task([_hh = _h] { delete _hh; });
+      });
+    } else {
+      _rel_func(_h);
+      delete _h;
+    }
+  }
+  handle_t get_handle() { return *_h; }
+  void add_dependency(sycl::event e) { _deps.push_back(e); }
+
+private:
+  handle_t *_h;
+  sycl::queue *_q;
+  init_func_t _init_func;
```

Comment:
We can save the 2 function ptr here, then we can get rid of 2nd and 3rd temp args.
We can try to use 2 static member function, for 2 speclizations, set different functions.
If there is redefinitaion issue, we can use them as general members.